### ISSUE TITLE
Silo will not start with Azure Event Hub stream provider and no Internet Access

### ISSUE THREAD

**Issue by kehigginbotham** [dotnet/orleans#6463]
We have a use case where our application must be able to run successfully in an environment with no Internet access.  We have accomplished this requirement with the exception of the startup of the silo.  When starting the silo, the following exception is throw:

Running in Windows Server environment, .NET Core 3.1.1

`System.Net.Sockets.SocketException (11001): No such host is known.
at Microsoft.Azure.EventHubs.Amqp.AmqpEventHubClient.CreateConnectionAsync(TimeSpan timeout)
at Microsoft.Azure.Amqp.FaultTolerantAmqpObject1.OnCreateAsync(TimeSpan timeout) at Microsoft.Azure.Amqp.Singleton1.GetOrCreateAsync(TimeSpan timeout)
at Microsoft.Azure.Amqp.Singleton1.GetOrCreateAsync(TimeSpan timeout) at Microsoft.Azure.EventHubs.Amqp.Management.AmqpServiceClient.OpenRequestResponseLinkAsync(String type, String address, Nullable1 entityType, String[] requiredClaims, TimeSpan timeout)
at Microsoft.Azure.EventHubs.Amqp.Management.AmqpServiceClient.OpenLinkAsync(TimeSpan timeout)
at Microsoft.Azure.Amqp.FaultTolerantAmqpObject1.OnCreateAsync(TimeSpan timeout) at Microsoft.Azure.Amqp.Singleton1.GetOrCreateAsync(TimeSpan timeout)
at Microsoft.Azure.Amqp.Singleton`1.GetOrCreateAsync(TimeSpan timeout)
at Microsoft.Azure.EventHubs.Amqp.Management.AmqpServiceClient.GetRuntimeInformationAsync()
at Microsoft.Azure.EventHubs.EventHubClient.GetRuntimeInformationAsync()
at Orleans.ServiceBus.Providers.EventHubAdapterFactory.GetPartitionIdsAsync()
at Orleans.ServiceBus.Providers.EventHubAdapterFactory.CreateAdapter()
at Orleans.Providers.Streams.Common.PersistentStreamProvider.Init(CancellationToken token)
at Orleans.Runtime.SiloLifecycleSubject.MonitoredObserver.OnStart(CancellationToken token)
at Orleans.LifecycleSubject.<OnStart>g__CallOnStart|7_0(Int32 stage, OrderedObserver observer, CancellationToken cancellationToken)
at Orleans.LifecycleSubject.OnStart(CancellationToken ct)
at Orleans.Runtime.Scheduler.AsyncClosureWorkItem.Execute()
at Orleans.Runtime.Silo.StartAsync(CancellationToken cancellationToken)`

**Comment by ReubenBond** [comment]
How would Azure Event Hubs work without Internet access? Is this running in an environment with direct access to Event Hubs?

This seems more like an Azure Event Hubs issue (I'm not sure what Orleans can do about it)

**Comment by KevinCathcart** [comment]
Some relevant quotes (of mine) from the gitter:

> Hmm... my (possibly incorrect) understanding is that the intention is for most stream providers to "work" in that scenario, but obviously while the Internet (or Azure) remains unavailable all attempts to send events on that provider would fail, and no events would be received on that provider. That [appears to be] the whole idea behind [the pulling agent] ignoring any exceptions thrown by the `IQueueAdapterReciever.Initialize` method. Once connectivity is restored, streaming on that provider is supposed to start working again.

> Coming back and looking closely at that stack trace, and this looks to be more of a design limitation of the EventHubs provider. On startup it queries to get the list of partition IDs from azure, in order to initialize the StreamQueueMapper. Unfortunately, that really does need to be eagerly initialized, since the QueueBalancer component uses it.
> I don't see any obvious simple change to the provider that would fix this, other than allowing you to configure a list of the partition ids manually, which would remove the need to check in with azure on boot. (In practice the ids are just the numbers from 0 to the partition count minus one, represented as strings.) That sounds like an entirely reasonable (and easy) feature to add to enable offline startup of a silo if Azure is not reachable at boot time.

I think being able to start silos without internet access (as long as the membership provider is still working) is something that Orleans ought to support in the inbox providers whenever feasible, with the system designed to recover once internet access is available again.

As mentioned in the last quote, it looks like allowing an optional manual specification of PartitionIds would enable that here.

**Comment by kehigginbotham** [comment]
@ReubenBond Our expectation is that they would not work without Internet access.  We utilize multiple stream providers simultaneously for redundancy purposes.  This was done to meet our business requirement of "no hard dependency on external services".

We have been using this approach successfully in production for some time.  However, our DR test failed because the DR environment "does not have Internet access" (or so I have been told by the network engineers), and the test performers were unable to start the silos.

**Comment by ReubenBond** [comment]
> Our expectation is that Event Hubs would not work without Internet access.

As far as I understand it, that expectation is being met (the silo terminates).
Is there any action which Orleans should take here?

**Comment by kehigginbotham** [comment]
While we expect Event Hubs would not work without Internet access, we do expect that Orleans would start the silo properly, even if the Event Hubs are unreachable.  In that scenario, our fall-back "secondary" stream provider carries the load.

If we have to restart the silos to connect to Event Hubs after they are reachable again, that is acceptable.

**Comment by ReubenBond** [comment]
In this case, though, the silo will never have connectivity (it's intentionally denied in that environment). If the silo starts "successfully" under those circumstances, it is potentially masking configuration errors.

I'm not dogmatic on the stance, though: I defer to others to weigh in. In particular, do you have thoughts on this, @jason-bragg?

**Comment by kehigginbotham** [comment]
The better approach would be for us to load our draconian DR test environment with a configuration that doesn't try to register the Azure stream provider.  Our secondary stream provider continues to carry the load, and everything should "just work" (albeit in a reduced capacity).

I'll leave the issue open to facilitate discussion, but I think we have a way around this "issue".  Thanks @ReubenBond and @KevinCathcart.

**Comment by jason-bragg** [comment]
Our provider model, in general, is intended to prevent silo startup if there are issues with any dependencies.  The idea is that it's better to not start than to start in a state which can't perform as expected.

With that in mind, the silo can be started with the stream providers not actively reading from eventhub until they are told to, but the initialization logic still expects to get the provider in a state which is ready to start reading (on command) which, in the case of EventHub, means connecting to the hub.

One, quite questionable, potential solution is to configure the stream lifecycle options to not initialize (or start) the stream provider until much later in the orleans lifecycle (maybe SiloLifecycleStages.Last - 1000?) and add a startup task before that (respectively SiloLifecycleStages.Last - 1010?) which does not resolve until it can talk to the internet.  The lifecycle will block on a stage until all work for that stage is completed, so when the silo hits the 'internet connected' stage and waits there until connection is found, then this would effectively delay the initialization and startup of the stream provider until after that condition is met.  There may be some startup timeout or some other issue with this which I'm unaware of, but I suspect it's worth trying, as I doubt it's much work.

It should be noted that the task returned from the silo host's StartAsync will not resolve until the lifecycle is fully started, so if there is work waiting on this, it will not be performed until the internet connection is established and the stream provider is started.  If that's the case, it may be necessary to start that work despite the fact that the silo is still starting (StartAsync Task not resolved)...?

In case it's unclear, the above suggestion is probably a bad idea.. but it's the best I can come up with at the moment.. :/

**Comment by KevinCathcart** [comment]
The Azure queue, GCP, and AWS streaming providers don’t actually prevent the silo from starting up if the service is unreachable. Therefore among the persistent streaming providers in this repository, EventHubs is a bit of the odd one out. Admittedly, this is hardly the only way in which it is different from the others.

On the other hand, streaming providers are a bit of the odd one out amongst provider types, as Clustering, Reminder, and Storage providers will block silo startup if the services are down.

> The idea is that it's better to not start than to start in a state which can't perform as expected.

The counter argument here is that in some cases it can be rather useful to be able to start a silo in a degraded state (equivalent to the state when the service in question has gone down after the silo already started), especially when a certain provider is not actually critical to the core functionality, but only impacts some non-critical side functionality. Let’s consider two scenarios in which this would be helpful.

The first scenario is deploying an important code fix while some non-critical provider’s backing service is experiencing downtime. The only real option for dealing with that case is to configure the silo to not add that provider for now, deploy it, wait for the outage to end, reconfigure the silo to add that provider again, and redeploy it. 

The second case is auto scaling. For an Orleans-based system that uses an autoscaler, if such a non-critical provider is down when prime usage hours begins, then the autoscaler would try to add more silos as load grows, but they would fail to start. Eventually the load will likely overwhelm the existing silos, and the result is that an outage of some non-critical component has cascaded to an outage of the entire system.

Once again this is possible to avoid by having the silo configuring code check if the non-critical provider’s backing service is working, and if not, to skip adding the provider to the system. Thus it would start up as a mostly-functional silo. Ideally it would have some background task to monitor the backing service in question, and when it becomes available again, to coordinate with the auto-scaler to either restart the silo so it can become fully functional, or to start a new replacement silo, and then stop the current one one.

Now do note that I don’t actually care very strongly about this. As I discussed above it is certainly possible to work around such a limitation. But having the option to start in a degraded state certainly could prove useful in some scenarios. But it may not be worthwhile to try to actually support this, especially since workarounds are possible, even if not ideal.

**Comment by ReubenBond** [comment]
> it may not be worthwhile to try to actually support this 

I can see how being able to start in a degraded state can cause improved behavior in some cases (when a catastrophe is occurring), but it also causes significantly worse behavior in other cases (when something has been misconfigured). Keeping things on the balance, I don't see this as something which is in need of change just yet, and I would prefer to avoid an explosion of options for things like starting up in a degraded state - we are even considering removing some such options which currently exist (`ClusterMembershipOptions.ValidateInitialConnectivity`) because they cause more harm than they remediate.

It's possible that I am misunderstanding something here, and that this is something we really should be changing or there's some way to ameliorate the downsides of taking such a change without increasing complexity and the configuration surface. If that's the case and someone feels strongly about this, please do open a new issue which references this one.
